esp/platform/espprotocol.cpp

```diff
@@ -755,3 +755,92 @@ int CEspProtocol::countBindings(int port)
     else
         return apport->getBindingCount();
 }
+
+//Class to detect if connection is closed or reset by peer
+class ConnectionCloseDetector : implements IConnectionCloseDetector, public CInterface
+{
+private:
+    class CloseDetectionThread : public Thread
+    {
+    private:
+        ISocket* m_sock;
+        bool m_stop = false;
+        ConnectionCloseDetector* m_detector;
+    public:
+        CloseDetectionThread(ISocket* sock, ConnectionCloseDetector* detector) : m_sock(sock), m_detector(detector) {}
+        void stop()
+        {
+            m_stop = true;
+            join();
+        }
+        virtual int run()
+        {
+            while (!m_stop)
+            {
+                int ret = m_sock->wait_read(100);
+                if (m_stop)
+                    break;
+                if (ret < 0)
+                {
+                    m_detector->notifyClose();
+                    break;
+                }
+                else if (ret > 0)
+                {
+                    if (m_sock->avail_read() == 0)
+                    {
+                        m_detector->notifyClose();
+                        break;
+                    }
+                    else
+                        Sleep(10); // avoid tight loop in case there're still outstanding data on the wire.
+                }
+            }
+            return 0;
+        }
+    };
+    CloseNotifyCallback m_callbackfunc = nullptr;
+    void* m_callbackdata = nullptr;
+    bool m_isPeerClosed = false;
```

Comment:
If one thread is reading m_isPeerClosed and another thread is writing m_isPerrClosed then maybe it should be an atomic ?  I mean if it gets updated slightly later perhaps its fine.

Reply:
I don't think atomic will buy us anything here, the reader will get either true or false, with or without atomic, depending on the timing of the call...

Reply:
True that the reader could've get a true with atomic but may get a false without, but that's a very, very rare occasion (writer calls first but finishes later), and even that happens it's no big deal, the reader will get the correct result next time it checks, and the callback function still sends back the notice instantaneously. Basically it's not worth the overhead to handle a rare harmless situation...

Reply:
Thats what I meant - as long as you are ok with how it could be then its fine.

Reply:
Yes this is ok.

Reply:
Same for stop ?  Is it ok or even possible to call m_detector->notifyClosed() from its dtor ?

Reply:
Yes it's the same for stop. I think it's better to call notifyClosed() right away instead of waiting for the destructor to avoid delay or complications.

Reply:
But its in a thread running in a loop ?